### DISCUSSION TITLE
multi site / multi market best approach

### DISCUSSION THREAD

**Discussion by simdoor** [pimcore/pimcore#11715] [Q&A / Support]
Hi everyone,
I would like to user Pimcore for a project where we'll have to handle multiple sites, each one having its own users, languages, subset of objects, etc. Kind of multi-tenancy.
I can see in the documentation that multiple sites are supported, but I fail to understand whether it offers more complex segregation than just at documents level.

For instance, say I have Product and Category objects, each site having its own set of items. How would I separate access to objects belonging to one site or the other? Using workspaces (for users) and folders could be a good approach ? Does it affect the results in the global admin search? If not, I would I guarantee separation when it comes to relationships, i.e. how can I make sure a Category can be added only to Products belonging to the same site? 

I was wondering whether there exists any solution out of the box or it should be coded.  

Any help and advice would be greatly appreciated! 
Thanks
Simone

**Comment by ckemptner** [comment]
Pimcore is not built for multi-tenancy. Most of our customers use the multi-site feature to create subsites/sub-domains for the same company and its sub-companies. So they typically share some of the content/data/products while some other data is only specific for certain countries or divisions. If you have completely different products and users, it will probably make sense to set up multiple dedicated tenants.

Your project sounds like it could make sense to use just one tenant because your sub-sites work with a sub-set of the data as far as I understand. In this case, you can make use of the various features Pimcore offers for restricting access to certain data or changing the views on data with workspaces (you already found it), custom perspectives, custom layouts (for objects), ...
Depending on the workspace configuration you have multiple options which will of course also affect the global admin search.

An out of the box solution for this topic does not exist because each company has different requirements. Pimcore only gives you the tooling to implement such solutions. You then have to do the customization by a mix of configuration and coding.

Regarding your question 

> I would guarantee separation when it comes to relationships, i.e. how can I make sure a Category can be added only to Products belonging to the same site?

Think about which settings the users want to / need to control in the data management (typically data objects) and which they need to control in the data output (e. .g. using documents and adding site-specific controls for certain contents inside the area bricks).

**Comment by simdoor** [comment]
Hi @ckemptner , 
thanks for your reply! Still have a few questions though, if you have the time to answer :)

_If you have completely different products and users, it will probably make sense to set up multiple dedicated tenants._
By this you mean multiple Pimcore instances  i.e applications running separately, right?

_Your project sounds like it could make sense to use just one tenant because your sub-sites work with a sub-set of the data as far as I understand._
That is correct, and I'd rather have a single application since all the business logic is shared across all sites. 
However, imagine I have to return to the frontend a simple list of the products available for the current site. 
What would be the best approach to "scope" the Object Listing to the specific site/market in your experience ? I am thinking about using the o_path attribute to query for a subset of data and find a way to automatically apply the same criteria to all objects listing (Product, Category, News, etc.). On the admin side, I would configure it in a way that users can only work within the site/market specific folder. Does this sound a sensible approach or are there known downsides to it?

**Reply by ckemptner** [reply]
> By this you mean multiple Pimcore instances i.e applications running separately, right?

Yes!

> What would be the best approach to "scope" the Object Listing to the specific site/market in your experience ? I am thinking about using the o_path attribute to query for a subset of data and find a way to automatically apply the same criteria to all objects listing (Product, Category, News, etc.). On the admin side, I would configure it in a way that users can only work within the site/market specific folder. Does this sound a sensible approach or are there known downsides to it?

It is easiest is if you can connect it to some rule based on your data. For example specific countries/languages/brands/categories/...
Or does it need to be configurable? Then add a checkbox to allow the users to release data to specific sites. A site-specific folder can be a good idea for data that has no hierarchies and has no connection with each other. When you have hierarchies and the same hierarchy is shared between markets you probably will want your users to work in the same folders and restrict the rights by using a combination of workspaces and custom layouts. The same applies if you have multiple languages for the same data (blog article for example).
